tests/testthat/test-go.R

```diff
@@ -1,21 +1,181 @@
 
 # Make a small GO table of the expected format
-go.df <- data.frame(
+rbp_df <- data.frame(
   UNIPROTKB = c("Q9BYJ9", "Q9NWX6"),
-  GO.ID = rbp.go
+  GO.ID = rbp_terms
 )
 
 #### Tests ---------------------------------------------------------------------
 
+test_that("determine_ancestor_function works", {
+  go_relations <- mapply(function(gt, ont) {
+    AnnotationDbi::get(
+      gt, determine_ancestor_function(gt, ont)
+    )
+  },
+  go_terms, onts)
+
+  expect_named(go_relations, go_terms)
+  expect_equivalent(lengths(go_relations), c(23, 9, 3))
+})
+
+test_that("determine_offspring_function works", {
+  go_relations <- mapply(function(gt, ont) {
+    AnnotationDbi::get(
+      gt, determine_offspring_function(gt, ont)
+    )
+  },
+  go_terms, onts)
+
+  expect_named(go_relations, go_terms)
+  expect_equivalent(lengths(go_relations), c(7, 14, 47))
+})
+
+test_that("get_all_mappings works with ancestor direction", {
+  rbp_mappings <- get_all_mappings(
+    go_ids = rbp_terms,
+    ontologies = rbp_ontologies,
+    verbose = TRUE,
+    direction = "ancestor"
+  )
+  expect_named(rbp_mappings, rbp_terms)
+  expect_equivalent(lengths(rbp_mappings), c(6, 7))
+})
+
+test_that("get_all_mappings works with offspring direction", {
+  rbp_mappings <- get_all_mappings(
+    go_ids = rbp_terms,
+    ontologies = rbp_ontologies,
+    verbose = TRUE,
+    direction = "offspring"
+  )
+
+  expect_named(rbp_mappings, rbp_terms)
+  expect_equivalent(lengths(rbp_mappings), c(173, 4))
+})
+
+test_that("expand_terms works", {
+  rbp_mappings <- get_all_mappings(
+    go_ids = rbp_terms,
+    ontologies = rbp_ontologies,
+    verbose = TRUE,
+    direction = "offspring"
+  )
+
+  rbp_expanded <- expand_terms(
+    go_df = rbp_df,
+    go_col = "GO.ID",
+    go2Ancestor = rbp_mappings
+  )
+
+  expect_equal(
+    rbp_expanded$GOID,
+    unique(c("GO:0003723", rbp_mappings[[1]], rbp_mappings[[2]]))
+  )
+})
+
 test_that("get_ancestor_go works", {
-  ancestor.terms <- get_ancestor_go(go.df, verbose = FALSE)
+  ancestor_terms <- get_ancestor_go(rbp_df, verbose = FALSE)
+
+  expect_equal(nrow(ancestor_terms), 13)
+  expect_false(any(is.na(ancestor_terms)))
+})
+
+test_that("get_ancestor_go early debug works", {
+  ancestor_terms <- get_ancestor_go(rbp_df, verbose = FALSE, return_early_debug = TRUE)
+
+  expect_identical(ancestor_terms[[1]], rbp_df)
+  expect_identical(
+    ancestor_terms[[2]],
+    get_all_mappings(
+      go_ids = rbp_terms,
+      ontologies = rbp_ontologies,
+      verbose = TRUE,
+      direction = "ancestor"
+    )
+  )
+})
+
+#### Sanity checks -------------------------------------------------------------
+
+test_that("determine_ancestor_function produces NA is ontology is NA", {
+  result <- determine_ancestor_function(rbp_terms[1], NA)
+  expect_identical(result, NA)
+})
+
+test_that("determine_ancestor_function produces NA is ontology is nonsense", {
+  result <- determine_ancestor_function(rbp_terms[1], "banana")
```

Comment:
Someone's been having too much fun testing :banana:

Reply:
I think you'll find some more fruit names scattered here and there 😆